src/exoticatechnologies/util/datastructures/tests/SynchronizedRingBufferTest.kt

```diff
@@ -0,0 +1,300 @@
+        assert(ring.get() == 2)
+
+        val test = ring.find { value -> value == 15 }
+
+        assert(test != null)
+        assert(test == 15)
+
+        assert(ring.get() == 3)
+        assert(ring.get() == 4)
+    }
+
+    @Test
+    fun after_capacity_is_rolled_over_THEN_get_still_returns_oldest_result() {
+        val ring = SynchronizedRingBuffer<Int>(10, -1, Int::class.java)
+        ring.setWaitingTimeoutMillis(100)
+
+        for (i in 1..10) {
+            ring.put(i)
+        }
+        // Oldest is 1,2,3
+
+        // put 20, rolling over 1
+        ring.put(20)
+
+        //oldest is 2,3,4..10, then 20
+        for (i in 2..10) assert(ring.get() == i)
+        assert(ring.get() == 20)
+    }
+
+    /**
+     * Waiting tests
+     */
+    @Test(timeout = 1000)
+    fun waiting_order_from_javadoc_is_respected_WHEN_testing_empty() {
+        val ring = SynchronizedRingBuffer<Int>(10, -1, Int::class.java)
+        ring.setWaitingTimeoutMillis(1000000)
+        ring.setNotEmptyTimeoutMillis(0)
+
+        assert(ring.get() == -1)
+    }
+
+    @Test
+    fun waiting_order_from_javadoc_is_respected_WHEN_testing_full() {
+        val ring = SynchronizedRingBuffer<Int>(10, -1, Int::class.java)
+        ring.setWaitingTimeoutMillis(1000000)
+        ring.setRollOverTimeoutMillis(0)
+
+        for (i in 1..10) {
+            ring.put(i)
+        }
+        // Oldest is 1,2,3
+
+        // put 20, rolling over 1
+        ring.put(20)
+
+        //oldest is 2,3,4..10, then 20
+        for (i in 2..10) assert(ring.get() == i)
+        assert(ring.get() == 20)
+    }
+
+    /**
+     * Multithreaded tests
+     */
+
+    @Test
+    fun multithreaded_test1_producer_and_consumer() {
+        val buffer = SynchronizedRingBuffer<Int>(10, -1, Int::class.java)
+        buffer.setWaitingTimeoutMillis(2000)
+        buffer.setRollOverTimeoutMillis(2000)
+
+        // Producer thread
+        val producerThread = Thread {
+            for (i in 0 until 10) {
+                buffer.put(i)
+                println("Produced: $i")
+            }
+        }
+
+        // Consumer thread
+        val consumerThread = Thread {
+            for (i in 0 until 10) {
+                val item = buffer.get()
+                println("Consumed: $item")
+            }
+        }
+
+        // Searching for an item - this part is flaky
+        // there is *very* little chances of starting this thread just when the producer has put it in
+        // and before the consumer has pulled it out, because searching just searches through the
+        // current state at time of calling. Nothing can be produced nor consumed while search is ongoing.
+//        var foundItem: Int? = null
+//        val searchThread = Thread {
+//            foundItem = buffer.find { it == 3 }
+//            if (foundItem != null) {
+//                println("Found item: $foundItem")
+//            } else {
+//                println("Item not found")
+//            }
+//        }
```

Comment:
would be worthwhile to get rid of this, but then again - it's a good reminder to not mix this into the test again, and just let it have it's own test

Reply:
oh no, this is the producer/consumer test. let this stay. resolving.